client.go

```diff
@@ -52,37 +52,66 @@ type ClientOption interface {
 	IsClientOption()
 }
 
-// NewClient makes a new Client (with default timeout)
-func NewClient(skipVerify bool, options ...ClientOption) *Client {
-	return NewClientWithTransportTimeout(requestTimeout, newFederationTripper(skipVerify, options...))
-}
-
-// NewClientWithTieout makes a new Client (with specified timeout)
-func NewClientWithTimeout(timeout time.Duration, skipVerify bool, options ...ClientOption) *Client {
-	return NewClientWithTransportTimeout(timeout, newFederationTripper(skipVerify, options...))
-}
-
-// NewClientWithTransport makes a new Client with an existing transport
-func NewClientWithTransport(transport http.RoundTripper) *Client {
-	return NewClientWithTransportTimeout(requestTimeout, transport)
-}
-
-// NewClientWithTransportTimeout makes a new Client with a specified request timeout
-func NewClientWithTransportTimeout(timeout time.Duration, transport http.RoundTripper) *Client {
-	return &Client{
+// NewClient makes a new Client. You can supply zero or more ClientOpts
+// which control the transport, timeout, TLS validation etc - see
+// WithTransport, WithTimeout, WithSkipVerify, WithDNSCache etc.
+func NewClient(options ...ClientOption) *Client {
+	var transport http.RoundTripper
+	var dnsCache *DNSCache
+	timeout := requestTimeout
+	skipVerify := false
+	for _, option := range options {
+		switch o := option.(type) {
+		case WithTransport:
+			transport = o.Transport
```

Comment:
Can we not do this please? Can we just use `WithTransport` the Go way as per https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis tl;dr

```go
type ClientOption func(*Client)

func NewClient(options ...ClientOption) *Client {
    cli := &Client{}
    for _, opt := range options {
        opt(cli)
    }
}

func WithTransport(tr http.Transport) ClientOption {
    return func(cli *Client) {
        cli.Transport = tr
    }
}
```

Reply:
Well, I can shuffle it around to use functions instead of structs, that's not difficult, but we'll still need to write to an options struct before building the `&Client`/`http.Client` because `http.RoundTripper` is an opaque type, and we need to not be sensitive to the ordering of options.